File: pgpainless-sop/src/main/java/org/pgpainless/sop/VersionImpl.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.sop;

import java.io.IOException;
import java.io.InputStream;
import java.util.Locale;
import java.util.Properties;

import org.bouncycastle.jce.provider.BouncyCastleProvider;
import sop.operation.Version;

/**
 * Implementation of the <pre>version</pre> operation using PGPainless.
 */
public class VersionImpl implements Version {

    // draft version
    private static final int SOP_VERSION = 7;

    @Override
    public String getName() {
        return "PGPainless-SOP";
    }

    @Override
    public String getVersion() {
        // See https://stackoverflow.com/a/50119235
        String version;
        try {
            Properties properties = new Properties();
            InputStream propertiesFileIn = getClass().getResourceAsStream("/version.properties");
            if (propertiesFileIn == null) {
                throw new IOException("File version.properties not found.");
            }
            properties.load(propertiesFileIn);
            version = properties.getProperty("version");
        } catch (IOException e) {
            version = "DEVELOPMENT";
        }
        return version;
    }

    @Override
    public String getBackendVersion() {
        return "PGPainless " + getVersion();
    }

    @Override
    public String getExtendedVersion() {
        double bcVersion = new BouncyCastleProvider().getVersion();
        String FORMAT_VERSION = String.format("%02d", SOP_VERSION);
        return getName() + " " + getVersion() + "\n" +
                "https://codeberg.org/PGPainless/pgpainless/src/branch/master/pgpainless-sop\n" +
                "\n" +
                "Implementation of the Stateless OpenPGP Protocol Version " + FORMAT_VERSION + "\n" +
                "https://datatracker.ietf.org/doc/html/draft-dkg-openpgp-stateless-cli-" + FORMAT_VERSION + "\n" +
                "\n" +
                "Based on pgpainless-core " + getVersion() + "\n" +
                "https://pgpainless.org\n" +
                "\n" +
                "Using " + String.format(Locale.US, "Bouncy Castle %.2f", bcVersion) + "\n" +
                "https://www.bouncycastle.org/java.html";
    }

    @Override
    public int getSopSpecRevisionNumber() {
        return SOP_VERSION;
    }

    @Override
    public boolean isSopSpecImplementationIncomplete() {
        return false;
    }

    @Override
    public String getSopSpecImplementationRemarks() {
        return null;
    }

}

File: pgpainless-sop/src/main/kotlin/org/pgpainless/sop/VersionImpl.kt
// SPDX-FileCopyrightText: 2024 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.sop

import java.io.IOException
import java.io.InputStream
import java.util.*
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.pgpainless.PGPainless
import sop.SOP
import sop.operation.Version

/** Implementation of the `version` operation using PGPainless. */
class VersionImpl(private val api: PGPainless) : Version {

    companion object {
        const val SOP_VERSION = 15
        const val SOPV_VERSION = "1.2"
    }

    override fun getBackendVersion(): String = "PGPainless ${getVersion()}"

    override fun getExtendedVersion(): String {
        val bcVersion =
            String.format(
                Locale.US, "Bouncy Castle %.2f", BouncyCastleProvider().versionStr.toDouble())
        val specVersion = String.format("%02d", SOP_VERSION)
        return """${getName()} ${getVersion()}
https://codeberg.org/PGPainless/pgpainless/src/branch/main/pgpainless-sop

Implementation of the Stateless OpenPGP Protocol Version $specVersion
https://datatracker.ietf.org/doc/html/draft-dkg-openpgp-stateless-cli-$specVersion

Based on pgpainless-core ${getVersion()}
https://pgpainless.org
${formatSopJavaVersion()}
Using $bcVersion
https://www.bouncycastle.org/java.html"""
    }

    override fun getName(): String = "PGPainless-SOP"

    override fun getSopSpecImplementationRemarks(): String? = null

    override fun getSopSpecRevisionNumber(): Int = SOP_VERSION

    override fun getSopVVersion(): String = SOPV_VERSION

    override fun getVersion(): String {
        // See https://stackoverflow.com/a/50119235
        return try {
            val resourceIn: InputStream =
                SOP::class.java.getResourceAsStream("/pgpainless-sop.properties")
                    ?: throw IOException("File pgpainless-sop.properties not found.")

            val properties = Properties().apply { load(resourceIn) }
            properties.getProperty("pgpainless-sop-version")
        } catch (e: IOException) {
            "DEVELOPMENT"
        }
    }

    private fun formatSopJavaVersion(): String {
        return getSopJavaVersion()?.let {
            """
            
            sop-java $it
            
        """
                .trimIndent()
        }
            ?: ""
    }

    override fun isSopSpecImplementationIncomplete(): Boolean = false
}
